aptos-move/framework/aptos-token/sources/bid.move

```diff
@@ -0,0 +1,281 @@
+        let coin = coin::withdraw<CoinType>(bidder, withdraw_amount);
+
+        let bid = Bid<CoinType> {
+            bidder: bidder_address,
+            coin,
+            token_id,
+            token_amount,
+            lock_until_sec,
+            listing_id,
+        };
+        let bid_id = create_bid_id(bidder);
+        initialize_bid_records<CoinType>(bidder);
+        let bid_records = &mut borrow_global_mut<BidRecords<CoinType>>(bidder_address).records;
+        table::add(bid_records, bid_id, bid);
+        bid_id
+    }
+
+    /// bidder can cancel a bid and get fund back
+    public entry fun cancel_bid<CoinType>(
+        bidder: &signer,
+        bid_id_creation_number: u64,
+    ) acquires BidRecords {
+        let bidder_address = signer::address_of(bidder);
+        let bid_id = guid::create_id(bidder_address, bid_id_creation_number);
+
+        let bid_records = &mut borrow_global_mut<BidRecords<CoinType>>(bidder_address).records;
+        assert!(table::contains(bid_records, bid_id), EBID_NOT_EXIST);
+
+        let bid = table::remove(bid_records, bid_id);
+        assert!(timestamp::now_seconds() >= bid.lock_until_sec, ECANNOT_DRAW_FUND_BEFORE_EXPIRATION_TIME);
+        coin::deposit(bidder_address, clear_bid(bid));
+    }
+
+    /// execute a bid to a listing, no signer required to perform this function
+    public fun execute_listing_bid<CoinType>(bidder: address, bid_id: ID, entry: Listing<CoinType>) acquires BidRecords {
+        let bid_records = &mut borrow_global_mut<BidRecords<CoinType>>(bidder).records;
+        assert!(table::contains(bid_records, bid_id), EBID_NOT_EXIST);
+        let bid = table::borrow(bid_records, bid_id);
+
+        let coin_owner = bid.bidder;
+        let token_owner = marketplace_utils::get_listing_owner(&entry);
+        // validate token_id match
+        assert!(bid.token_id == marketplace_utils::get_listing_token_id(&entry), ETOKEN_ID_NOT_MATCH);
+        // validate offerred amount and price
+        let listed_amount =  marketplace_utils::get_listing_token_amount(&entry);
+        let min_total = marketplace_utils::get_listing_min_price(&entry) * listed_amount;
+        assert!(coin::value(&bid.coin) >= min_total, ENO_SUFFICIENT_FUND);
+        assert!(bid.token_amount == listed_amount, ETOKEN_AMOUNT_NOT_MATCH);
+        // validate expiration time
+        let now = timestamp::now_seconds();
+        // only when lock time expires, funds can be extracted for executions
+        assert!(bid.lock_until_sec <= now, ECANNOT_DRAW_FUND_BEFORE_EXPIRATION_TIME);
+        // listing should expire after auction ends
+        assert!(marketplace_utils::get_listing_expiration(&entry) >= now, ELISTING_EXPIRED);
+        //listing_id matches
+        assert!(marketplace_utils::get_listing_id(&entry) == bid.listing_id, ELISTING_ID_NOT_MATCH);
+
+        // transfer coin and token
+        let token = token::withdraw_with_event_internal(token_owner, bid.token_id, listed_amount);
```

Comment:
I hava a doubt here, bidder locked the real coin for swap the token, but if seller don't have the token anymore, the auction will turn to failure.
I really think lock token in listing data is a better way to deal with that, if marketplace_utils is a standard contract, then wallet or other ecosystem project can checkout token ownership and it will have a good way to cancel all the listing or bidding when seller(token owner) withdraw token out of listing data.

Reply:
cc @areshand